File: scripts/test_aarch64_futex_wait_atomic.py
#!/usr/bin/env python3
"""Exercise the production futex read/enqueue critical section against a wake.

The kernel's scheduler-closure prefix through FutexTable::wait and its errno
mapping is used unchanged. A deterministic pre-lock scheduling hook supplies
the adversarial clear/wake ordering; the queue implementation is makos-futex.
This does not emulate interrupts or execute the guest scheduler's context swap.
"""

from pathlib import Path
import re
import subprocess
import tempfile

ROOT = Path(__file__).resolve().parents[1]
PROCESS = (ROOT / "kernel/src/aarch64_process.rs").read_text()


def item(source: str, start: str) -> str:
    offset = source.index(start)
    opening = source.index("{", offset)
    depth = 1
    end = opening + 1
    while depth:
        depth += (source[end] == "{") - (source[end] == "}")
        end += 1
    return source[offset:end]


futex = item(PROCESS, "pub fn futex(")
wait_path = futex[futex.index("if command != FUTEX_WAIT {"):]
opening = "let result = with_state(|state| {"
before, registration_and_rest = wait_path.split(opening, 1)
registration = registration_and_rest.split("state.contexts[index].context = captured;", 1)[0]
read = "let observed = unsafe { core::ptr::read_volatile(address as *const u32) };"
assert read not in before, "futex observation moved outside scheduler lock"
assert registration.count(read) == 1
assert registration.index(read) < registration.index("state.futex.wait(")
assert "with_state" not in registration
assert not re.sub(r"//[^\n]*", "", registration[registration.index(read) + len(read):
                                                       registration.index("let handle = match")]).strip()
assert "state.table.block_current_on(scheduler_cpu())" in registration_and_rest
assert "wake_futex_in_state(state, FutexKey::new(root, address), value as usize)" in futex
constants = "\n".join(re.findall(r"    const (?:EAGAIN|EINVAL|ETIMEDOUT):.*;$", futex, re.M))
errno = item(PROCESS, "const fn negative_errno(")
function_head = """
fn register_wait(address: u64, value: u32, now: u64, deadline: Option<u64>)
    -> Result<WaitHandle, u64> {
""" + constants + "\n"
tail = "\n        Ok(handle)\n    })\n}\n"
production = errno + function_head + "    with_state(|state| {\n" + registration + tail

with tempfile.TemporaryDirectory(prefix="makos-futex-wait-atomic-") as temporary:
    directory = Path(temporary)
    library = directory / "libmakos_futex.rlib"
    subprocess.run(["rustc", "--edition=2024", "--crate-name", "makos_futex", "--crate-type", "rlib",
                    str(ROOT / "crates/futex/src/lib.rs"), "-o", str(library)], check=True)
    source = directory / "registration.rs"
    source.write_text(production)
    (directory / "test.rs").write_text((ROOT / "scripts/test_aarch64_futex_wait_atomic.rs").read_text())

    def compile_fixture(name: str) -> Path:
        binary = directory / name
        subprocess.run(["rustc", "--edition=2024", "--test", str(directory / "test.rs"),
                        "--extern", f"makos_futex={library}", "-o", str(binary)], check=True)
        return binary

    subprocess.run([str(compile_fixture("futex-wait-atomic")), "--test-threads=1"],
                   check=True, timeout=20)
    # Move just the real volatile read before lock acquisition, recreating the
    # original ordering without altering the actual queue or value predicate.
    source.write_text(errno + function_head + read + "\n    with_state(|state| {\n"
                      + registration.replace(read, "") + tail)
    negative = subprocess.run([str(compile_fixture("futex-wait-before-lock")), "--exact",
                               "clear_and_empty_wake_before_lock_rejects_stale_wait"],
                              capture_output=True, text=True, timeout=10)
    if negative.returncode == 0 or "lost wake: stale futex sample queued after clear" not in negative.stdout + negative.stderr:
        raise SystemExit(f"pre-lock observation negative control did not expose lost wake:\n{negative.stdout}{negative.stderr}")

print("MAKOS_AARCH64_FUTEX_WAIT_ATOMIC_HOST_OK registration=production-code queue=makos-futex "
      "interleaving=clear-wake-before-lock result=EAGAIN read,enqueue=same-lock "
      "post-enqueue=woken roots=isolated errno=preserved negative_control=stale-wait-reproduced")

File: scripts/test_aarch64_futex_wait_atomic.rs
//! Deterministic adversarial scheduling point around real futex registration.
#![allow(dead_code)]

use makos_futex::{FutexKey, FutexTable, TaskId, WaitError, WaitHandle, WaitOutcome, WaiterState};
use std::sync::Mutex;
use std::sync::atomic::{AtomicU64, AtomicUsize, Ordering};

struct Table {
    current: Option<u64>,
}
impl Table {
    fn current_pid_on(&self, cpu: usize) -> Option<u64> {
        assert_eq!(cpu, 1);
        self.current
    }
}
struct Context {
    ttbr0: u64,
}
struct Slot {
    pid: u64,
    group_pid: u64,
    context: Context,
}
struct SchedulerState {
    table: Table,
    contexts: Vec<Slot>,
    futex: FutexTable<4>,
}
static STATE: Mutex<SchedulerState> = Mutex::new(SchedulerState {
    table: Table { current: None },
    contexts: Vec::new(),
    futex: FutexTable::new(),
});
static BEFORE_LOCK: AtomicUsize = AtomicUsize::new(0);
static WORD_ADDRESS: AtomicU64 = AtomicU64::new(0);
static LAST_WAKE_COUNT: AtomicUsize = AtomicUsize::new(usize::MAX);
static WAKE_CALLBACKS: AtomicUsize = AtomicUsize::new(0);
const ROOT: u64 = 0x4012_e000;

fn scheduler_cpu() -> usize {
    1
}

fn wake(key: FutexKey) -> usize {
    STATE
        .lock()
        .unwrap()
        .futex
        .wake(key, usize::MAX, |task, _| {
            assert_eq!(task, TaskId::new(7, 9));
            WAKE_CALLBACKS.fetch_add(1, Ordering::Relaxed);
        })
        .unwrap()
}

fn clear_and_wake() -> usize {
    let address = WORD_ADDRESS.load(Ordering::Relaxed);
    // As in clear-child-TID, publish zero before acquiring the futex/scheduler
    // lock for wake. Deterministic hooks order accesses without host races.
    unsafe { (address as *mut u32).write_volatile(0) };
    wake(FutexKey::new(ROOT, address))
}

fn with_state<R>(function: impl FnOnce(&mut SchedulerState) -> R) -> R {
    match BEFORE_LOCK.swap(0, Ordering::Relaxed) {
        0 => {}
        1 => {
            LAST_WAKE_COUNT.store(clear_and_wake(), Ordering::Relaxed);
        }
        2 => {
            LAST_WAKE_COUNT.store(
                wake(FutexKey::new(ROOT, WORD_ADDRESS.load(Ordering::Relaxed))),
                Ordering::Relaxed,
            );
        }
        _ => panic!("unknown scheduling hook"),
    }
    let mut state = STATE.lock().unwrap();
    function(&mut state)
}

include!("registration.rs");

fn setup(word: &mut u32) -> u64 {
    *STATE.lock().unwrap() = SchedulerState {
        table: Table { current: Some(9) },
        contexts: vec![Slot {
            pid: 9,
            group_pid: 7,
            context: Context { ttbr0: ROOT },
        }],
        futex: FutexTable::new(),
    };
    let address = word as *mut u32 as u64;
    WORD_ADDRESS.store(address, Ordering::Relaxed);
    BEFORE_LOCK.store(0, Ordering::Relaxed);
    LAST_WAKE_COUNT.store(usize::MAX, Ordering::Relaxed);
    WAKE_CALLBACKS.store(0, Ordering::Relaxed);
    address
}

#[test]
fn clear_and_empty_wake_before_lock_rejects_stale_wait() {
    for deadline in [None, Some(500)] {
        let mut word = 9;
        let address = setup(&mut word);
        BEFORE_LOCK.store(1, Ordering::Relaxed);
        let result = register_wait(address, 9, 100, deadline);
        assert_eq!(word, 0);
        assert_eq!(LAST_WAKE_COUNT.load(Ordering::Relaxed), 0);
        assert_eq!(
            result,
            Err(negative_errno(11)),
            "lost wake: stale futex sample queued after clear"
        );
        let state = STATE.lock().unwrap();
        assert_eq!(state.futex.waiting(), 0);
        assert_eq!(state.futex.occupied(), 0);
    }
}

#[test]
fn enqueue_before_clear_is_woken_and_consumable() {
    let mut word = 9;
    let address = setup(&mut word);
    let handle = register_wait(address, 9, 100, None).unwrap();
    assert_eq!(
        STATE.lock().unwrap().futex.state(handle),
        Ok(WaiterState::Waiting)
    );
    assert_eq!(clear_and_wake(), 1);
    assert_eq!(word, 0);
    assert_eq!(WAKE_CALLBACKS.load(Ordering::Relaxed), 1);
    let mut state = STATE.lock().unwrap();
    assert_eq!(state.futex.waiting(), 0);
    assert_eq!(
        state.futex.take_outcome(handle),
        Ok(Some(WaitOutcome::Woken))
    );
    assert_eq!(state.futex.occupied(), 0);
}

#[test]
fn wake_without_value_change_does_not_forbid_a_new_wait() {
    let mut word = 9;
    let address = setup(&mut word);
    BEFORE_LOCK.store(2, Ordering::Relaxed);
    let handle = register_wait(address, 9, 100, None).unwrap();
    assert_eq!(LAST_WAKE_COUNT.load(Ordering::Relaxed), 0);
    assert_eq!(word, 9);
    assert_eq!(
        STATE.lock().unwrap().futex.state(handle),
        Ok(WaiterState::Waiting)
    );
    assert_eq!(clear_and_wake(), 1);
}

#[test]
fn wake_in_another_root_cannot_consume_this_wait() {
    let mut word = 9;
    let address = setup(&mut word);
    let handle = register_wait(address, 9, 100, None).unwrap();
    assert_eq!(wake(FutexKey::new(ROOT + 4096, address)), 0);
    assert_eq!(
        STATE.lock().unwrap().futex.state(handle),
        Ok(WaiterState::Waiting)
    );
    assert_eq!(wake(FutexKey::new(ROOT, address)), 1);
}

#[test]
fn existing_value_and_timeout_errno_are_preserved() {
    let mut word = 0;
    let address = setup(&mut word);
    assert_eq!(
        register_wait(address, 9, 100, None),
        Err(negative_errno(11))
    );
    // Value mismatch retains precedence over an already expired deadline.
    assert_eq!(
        register_wait(address, 9, 100, Some(99)),
        Err(negative_errno(11))
    );
    word = 9;
    assert_eq!(
        register_wait(address, 9, 100, Some(99)),
        Err(negative_errno(110))
    );
    assert_eq!(
        register_wait(address, 9, 100, Some(100)),
        Err(negative_errno(110))
    );
    assert_eq!(word, 9);
    assert_eq!(STATE.lock().unwrap().futex.waiting(), 0);
}

#[test]
fn absent_current_task_or_context_still_returns_invalid() {
    for missing_current in [true, false] {
        let mut word = 9;
        let address = setup(&mut word);
        {
            let mut state = STATE.lock().unwrap();
            if missing_current {
                state.table.current = None;
            } else {
                state.contexts.clear();
            }
        }
        assert_eq!(
            register_wait(address, 9, 100, None),
            Err(negative_errno(22))
        );
        assert_eq!(word, 9);
        assert_eq!(STATE.lock().unwrap().futex.waiting(), 0);
    }
}
